### ISSUE TITLE
Descriptor buffers size issues

### ISSUE THREAD

**Issue by Agrael1** [KhronosGroup/Vulkan-ValidationLayers#8203]
Hi, I have reverted to using bytes for descriptor tables. 

However there is a validation problem. If I bind and set offset 0 for a simple layout it starts making noisy messages in validation.
```
The offsets in pOffsets must be small enough such that any descriptor binding referenced by layout without the VK_DESCRIPTOR_BINDING_VARIABLE_DESCRIPTOR_COUNT_BIT flag computes a valid address inside the underlying VkBuffer
```

But if I change the size of the buffer to 256, the error goes away. I am using a huge buffer for bindless and bindful tests, so I may or may not be able to query the size of the descriptor layout. The layout itself does not exceed a single uniform binding, so allocating a buffer for 1 uniform should be enough. May this be a rule, that I missed, or is it a validation false positive?

**Comment by spencer-lunarg** [comment]
@Agrael1 I am still looking into this, what my "guess" is going on is the VVL is throwing a real error, but giving a terrible message (so needs to be fixed regardless)

If you call `vkGetDescriptorSetLayoutSizeEXT` on your layout, is it over `3 * (mutable_desc_size)`, if see the issue is not the offset, but that the offset `(0)` plus the `pLayoutSizeInBytes` is too large for your buffer

**Comment by Agrael1** [comment]
@spencer-lunarg The thing is, it is not registering as an error, VkResult is fine, and rendering is fine as well. `3 * (mutable_desc_size)` is enough to fit all descriptors, but pLayoutSizeInBytes is not giving adequate size. `256` is something worth mentioning in documentation of VK_EXT_descriptor_buffer, since it makes use of desc buffers challenging. I have tested it on several cards, none have any problems reading descriptors.

**Comment by spencer-lunarg** [comment]
>  but pLayoutSizeInBytes is not giving adequate size

Just to confirm, did you query the size... this is what the validation layers is currently using to decide (if that is correct or not, is something I am still trying to figure out)

**Comment by Agrael1** [comment]
yes, it returns 256.

**Comment by Agrael1** [comment]
`Create a VkBuffer, size equal to NumDescriptors multiplied by the descriptor size within it`

https://github.com/KhronosGroup/Vulkan-Docs/blob/main/proposals/VK_EXT_descriptor_buffer.adoc
Paragraph 4.1. 
Now it seems that the check is incorrect, and the buffer should be indeed `3 * (mutable_desc_size)`

**Comment by spencer-lunarg** [comment]
Still need to get back into Descriptor Buffers, but had this test lying around on my machine for when I (or whoever) get to it

```

TEST_F(PositiveDescriptorBuffer, BindingOffsetZero) {
    TEST_DESCRIPTION("https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/8203");
    SetTargetApiVersion(VK_API_VERSION_1_2);
    AddRequiredExtensions(VK_EXT_DESCRIPTOR_BUFFER_EXTENSION_NAME);
    AddRequiredExtensions(VK_KHR_MAINTENANCE_6_EXTENSION_NAME);
    AddRequiredFeature(vkt::Feature::bufferDeviceAddress);
    AddRequiredFeature(vkt::Feature::descriptorBuffer);
    RETURN_IF_SKIP(Init());

    VkPhysicalDeviceDescriptorBufferPropertiesEXT descriptor_buffer_properties = vku::InitStructHelper();
    GetPhysicalDeviceProperties2(descriptor_buffer_properties);

    const VkDescriptorSetLayoutBinding bindings[] = {
        {0, VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, 1, VK_SHADER_STAGE_VERTEX_BIT, nullptr},
    };
    const auto dslci1 = vku::InitStruct<VkDescriptorSetLayoutCreateInfo>(
        nullptr, static_cast<VkDescriptorSetLayoutCreateFlags>(VK_DESCRIPTOR_SET_LAYOUT_CREATE_DESCRIPTOR_BUFFER_BIT_EXT),
        size32(bindings), bindings);
    vkt::DescriptorSetLayout dsl1(*m_device, dslci1);

    VkPipelineLayoutCreateInfo plci = vku::InitStructHelper();
    plci.setLayoutCount = 1;
    plci.pSetLayouts = &dsl1.handle();
    vkt::PipelineLayout pipeline_layout(*m_device, plci);

    const VkDeviceSize buffer_size = 4 * descriptor_buffer_properties.descriptorBufferOffsetAlignment;

    VkBufferCreateInfo buffer_ci = vku::InitStructHelper();
    buffer_ci.size = buffer_size;
    buffer_ci.usage = VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT;

    VkMemoryAllocateFlagsInfo allocate_flag_info = vku::InitStructHelper();
    allocate_flag_info.flags = VK_MEMORY_ALLOCATE_DEVICE_ADDRESS_BIT;
    vkt::Buffer d_buffer(*m_device, buffer_ci, VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT, &allocate_flag_info);

    const VkDeviceAddress common_address = d_buffer.Address();

    VkDescriptorBufferBindingInfoEXT dbbi = vku::InitStructHelper();
    dbbi.address = common_address;
    dbbi.usage = VK_BUFFER_USAGE_SHADER_DEVICE_ADDRESS_BIT;

    m_command_buffer.Begin();
    vk::CmdBindDescriptorBuffersEXT(m_command_buffer.handle(), 1, &dbbi);

    constexpr uint32_t index = 0;
    const VkDeviceSize offset = 0;
    VkSetDescriptorBufferOffsetsInfoEXT set_info = vku::InitStructHelper();
    set_info.stageFlags = VK_SHADER_STAGE_ALL_GRAPHICS;
    set_info.layout = pipeline_layout.handle();
    set_info.firstSet = 0;
    set_info.setCount = 1;
    set_info.pBufferIndices = &index;
    set_info.pOffsets = &offset;

    vk::CmdSetDescriptorBufferOffsets2EXT(m_command_buffer.handle(), &set_info);

    m_command_buffer.End();
}
```

**Comment by spencer-lunarg** [comment]
@Agrael1 I think I fixed things in https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/10387 (will be out soon in the 1.4.321 SDK)

going to close this, if you find more descriptor buffer issues, please raise a new issue, I am MUCH more up-to-date with the code/spec now to fix things
